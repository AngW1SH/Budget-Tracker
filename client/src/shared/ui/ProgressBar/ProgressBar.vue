<script setup lang="ts">
defineProps<{
  progress: number;
  goal?: number;
}>();
</script>

<template>
  <div class="flex justify-between items-center">
    <div class="relative mt-px w-full h-2 bg-[#F4EBFF] rounded-lg">
      <div
        class="relative h-full bg-[#7F56D9] rounded-lg"
        :style="{ width: progress + '%' }"
      ></div>
      <div
        v-if="goal !== undefined"
        class="-top-[0.5rem] left-0 absolute rounded-lg h-6 w-[3px] bg-[#53389E]"
        :style="{ left: goal + '%' }"
      ></div>
    </div>
    <div class="w-max ml-[0.625rem] text-sm text-[#344054] font-medium">
      {{ progress + "%" }}
    </div>
  </div>
</template>
