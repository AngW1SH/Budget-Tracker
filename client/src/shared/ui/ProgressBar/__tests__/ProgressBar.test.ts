import { mount } from "@vue/test-utils";
import { describe, expect, test } from "vitest";
import ProgressBar from "../ProgressBar.vue";

describe("ProgressBar shared UI-component", () => {
  test("uses 'progress' prop to display percentage", () => {
    const progressBar = mount(ProgressBar, {
      props: {
        progress: 20,
      },
    });

    expect(progressBar.find("div:contains(20%)").exists()).toBe(true);
  });

  /*
  The next two tests are too implementation-specific,
  but I couldn't think of a way to select the goal-line somehow differently,
  because I really didn't want to use data-test attributes
  */
  test("displays the goal-line when the 'goal' prop is provided", () => {
    const progressBar = mount(ProgressBar, {
      props: {
        progress: 20,
        goal: 40,
      },
    });

    expect(progressBar.find(".absolute").exists()).toBe(true);
  });

  test("doesn't display the goal-line when the 'goal' prop is not provided", () => {
    const progressBar = mount(ProgressBar, {
      props: {
        progress: 20,
      },
    });

    expect(progressBar.find(".absolute").exists()).toBe(false);
  });

  test("snapshot UI", () => {
    const progressBar = mount(ProgressBar, {
      props: {
        progress: 20,
      },
    });
    expect(progressBar.element).toMatchSnapshot();
  });
});
