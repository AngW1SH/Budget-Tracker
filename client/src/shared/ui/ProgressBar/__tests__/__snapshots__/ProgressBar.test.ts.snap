// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressBar shared UI-component > snapshot UI 1`] = `
<div
  class="flex justify-between items-center"
>
  <div
    class="relative mt-px w-full h-2 bg-[#F4EBFF] rounded-lg"
  >
    <div
      class="relative h-full bg-[#7F56D9] rounded-lg"
      style="width: 20%;"
    />
    <!--v-if-->
  </div>
  <div
    class="w-max ml-[0.625rem] text-sm text-[#344054] font-medium"
  >
    20%
  </div>
</div>
`;
